In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/human-activity-recognition-with-smartphones/train.csv
/kaggle/input/human-activity-recognition-with-smartphones/test.csv


In [2]:
# import the necessary Libraries 
import numpy as np 
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# load the train dataset file 
df=pd.read_csv('/kaggle/input/human-activity-recognition-with-smartphones/train.csv')
df.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0         -0.923527         -0.934724  ...                        -0.710304   
1         -0.957686         -0.943068  ...                        -0.861499   
2         -0.977469         -0.938692  ...                        -0.760104   
3         -0.989302         -0.938692  ...                        -0.482845   
4         -0.990441         -0.942469  ...                        -0.699205   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                    -0.112754                              0.030400   
1                     0.053477                             -0.007435   
2                    -0.118559                              0.177899   
3                    -0.036788                             -0.012892   
4                     0.123320                              0.122542   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.464761                             -0.018446   
1                         -0.732626                              0.703511   
2                          0.100699                              0.808529   
3                          0.640011                             -0.485366   
4                          0.693578                             -0.615971   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  subject  \
0             -0.841247              0.179941             -0.058627        1   
1             -0.844788              0.180289             -0.054317        1   
2             -0.848933              0.180637             -0.049118        1   
3             -0.848649              0.181935             -0.047663        1   
4             -0.847865              0.185151             -0.043892        1   

   Activity  
0  STANDING  
1  STANDING  
2  STANDING  
3  STANDING  
4  STANDING  

[5 rows x 563 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 31.6+ MB


In [5]:
df.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject', 'Activity'],
      dtype='object', length=563)

In [7]:
df.shape

(7352, 563)

**Here we see that we have toatal 563 features along with one target feature Activity**

In [6]:
df['Activity'].value_counts()

Activity
LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: count, dtype: int64

In [18]:
# load the test dataset file  
df_test=pd.read_csv('/kaggle/input/human-activity-recognition-with-smartphones/train.csv')
df_test.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0         -0.923527         -0.934724  ...                        -0.710304   
1         -0.957686         -0.943068  ...                        -0.861499   
2         -0.977469         -0.938692  ...                        -0.760104   
3         -0.989302         -0.938692  ...                        -0.482845   
4         -0.990441         -0.942469  ...                        -0.699205   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                    -0.112754                              0.030400   
1                     0.053477                             -0.007435   
2                    -0.118559                              0.177899   
3                    -0.036788                             -0.012892   
4                     0.123320                              0.122542   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.464761                             -0.018446   
1                         -0.732626                              0.703511   
2                          0.100699                              0.808529   
3                          0.640011                             -0.485366   
4                          0.693578                             -0.615971   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  subject  \
0             -0.841247              0.179941             -0.058627        1   
1             -0.844788              0.180289             -0.054317        1   
2             -0.848933              0.180637             -0.049118        1   
3             -0.848649              0.181935             -0.047663        1   
4             -0.847865              0.185151             -0.043892        1   

   Activity  
0  STANDING  
1  STANDING  
2  STANDING  
3  STANDING  
4  STANDING  

[5 rows x 563 columns]

In [91]:
df_test.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-skewness()', 'fBodyBodyGyroJerkMag-kurtosis()',
       'angle(tBodyAccMean,gravity)', 'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject'],
      dtype='object', length=562)

In [23]:
# seperate target column for test data 

X_test=df.drop(columns=['Activity'])
y_test=df['Activity']


In [24]:
X_test.shape

(7352, 562)

In [11]:
#seperate one target column  for train dataset 
X_train=df.drop(columns=['Activity'])

y_train=df['Activity']

In [14]:
X_train.shape

(7352, 562)

In [13]:
y_train.shape

(7352,)

**we are simply use knn Classifier**

In [16]:
from sklearn.neighbors import KNeighborsClassifier

knn=KNeighborsClassifier()

knn.fit(X_train,y_train)

KNeighborsClassifier()

In [25]:
y_pred=knn.predict(X_test)

In [65]:
from sklearn.metrics import *

print("Accuracy of knn model with all features ",accuracy_score(y_test,y_pred)*100)

Accuracy of knn model with all features  98.89825897714908


**WITH PCA**

In [84]:
from sklearn.decomposition import PCA

pca= PCA(n_components=2)

In [85]:
X_train_pca=pca.fit_transform(X_train)
X_test_pca=pca.transform(X_test)

X_train_pca.shape

(7352, 2)

In [86]:
X_test_pca.shape

(7352, 2)

In [75]:
X_train.shape

(7352, 562)

Here we see that after applying PCA the shape of the dataset or dimension are reduced upto 2.

In [82]:
%matplotlib inline

In [87]:
import plotly.express as px

y_train_pca=y_train.astype(str)

fig= px.scatter(    x=X_train_pca[:,0],
                    y=X_train_pca[:,1],
                    color=y_train_pca,
                    color_discrete_sequence=px.colors.qualitative.G10)
fig.show()

In this we see that all categories points but the some of points are hidden in 2d pot so we cant differentiate , let's make us 3D and Flexible 

In [89]:
pca= PCA(n_components=3)

X_train_pca=pca.fit_transform(X_train)
X_test_pca=pca.transform(X_test)


In [90]:
y_train_pca=y_train.astype(str)

fig= px.scatter_3d(    x=X_train_pca[:,0],
                    y=X_train_pca[:,1],
                    z=X_train_pca[:,2],
                    color=y_train_pca )
fig.update_layout(
           margin=dict(l=20, r=20, t=20, b=20)
)
                    
fig.show()

Here is the best representation in 3D for three principal components...

In [76]:
knn=KNeighborsClassifier()

In [77]:
knn.fit(X_train_pca,y_train)

KNeighborsClassifier()

In [78]:
y_pred_pca=knn.predict(X_test_pca)

In [79]:
print("Accuracy of knn model with Principal components ",accuracy_score(y_test,y_pred_pca))

Accuracy of knn model with Principal components  0.9094124047878128


In [62]:
for i in range(20):
    pca= PCA(n_components=i+1) # because of first value is 0

    X_train_pca=pca.fit_transform(X_train)
    X_test_pca=pca.transform(X_test)

    knn=KNeighborsClassifier()
    knn.fit(X_train_pca,y_train)
    
    y_pred_pca=knn.predict(X_test_pca)
    
    print(f"Accuracy of knn model with {i} PC :",accuracy_score(y_test,y_pred_pca))

Accuracy of knn model with 0 PC : 0.6516594124047879
Accuracy of knn model with 1 PC : 0.7417029379760609
Accuracy of knn model with 2 PC : 0.8053590859630033
Accuracy of knn model with 3 PC : 0.9094124047878128
Accuracy of knn model with 4 PC : 0.9294069640914037
Accuracy of knn model with 5 PC : 0.9474972796517954
Accuracy of knn model with 6 PC : 0.9532100108813928
Accuracy of knn model with 7 PC : 0.9563384113166485
Accuracy of knn model with 8 PC : 0.954842219804135
Accuracy of knn model with 9 PC : 0.9602829162132753
Accuracy of knn model with 10 PC : 0.9616430903155604
Accuracy of knn model with 11 PC : 0.9640914036996736
Accuracy of knn model with 12 PC : 0.9673558215451578
Accuracy of knn model with 13 PC : 0.9707562568008705
Accuracy of knn model with 14 PC : 0.9674918389553863
Accuracy of knn model with 15 PC : 0.9691240478781284
Accuracy of knn model with 16 PC : 0.9691240478781284
Accuracy of knn model with 17 PC : 0.9699401523394995
Accuracy of knn model with 18 PC : 0.97

Here we see that if we increase the number of principal components then accuracy of model is also increase.

So if we compare the accuracy of  knn model that include all features and this knn model that uses principal components we get the approximately same accuracy score..

And may be if we calculate with more pc we reach to exact same accuracy 

Let's do this :

In [64]:
for i in range(40):
    pca= PCA(n_components=i+1) # because of first value is 0

    X_train_pca=pca.fit_transform(X_train)
    X_test_pca=pca.transform(X_test)

    knn=KNeighborsClassifier()
    knn.fit(X_train_pca,y_train)
    
    y_pred_pca=knn.predict(X_test_pca)
    
    print(f"Accuracy of knn model with {i} PC :",accuracy_score(y_test,y_pred_pca)*100)

Accuracy of knn model with 0 PC : 65.16594124047879
Accuracy of knn model with 1 PC : 74.17029379760609
Accuracy of knn model with 2 PC : 80.53590859630033
Accuracy of knn model with 3 PC : 90.94124047878128
Accuracy of knn model with 4 PC : 92.94069640914037
Accuracy of knn model with 5 PC : 94.74972796517955
Accuracy of knn model with 6 PC : 95.32100108813928
Accuracy of knn model with 7 PC : 95.6474428726877
Accuracy of knn model with 8 PC : 95.49782372143635
Accuracy of knn model with 9 PC : 96.02829162132753
Accuracy of knn model with 10 PC : 96.19151251360174
Accuracy of knn model with 11 PC : 96.40914036996736
Accuracy of knn model with 12 PC : 96.74918389553862
Accuracy of knn model with 13 PC : 97.0076169749728
Accuracy of knn model with 14 PC : 96.77638737758433
Accuracy of knn model with 15 PC : 96.89880304679
Accuracy of knn model with 16 PC : 96.89880304679
Accuracy of knn model with 17 PC : 96.96681175190425
Accuracy of knn model with 18 PC : 97.04842219804135
Accuracy of

In [ ]:
# as we have mentioned above we get the same accuracy score -- 98 %